### DISCUSSION TITLE
Testing within directus

### DISCUSSION THREAD

**Discussion by mayankpunetha007** [directus/directus#10607] [Archive]
Is there some guidelines or some examples to follow to test the action/filters in my directus application, Few things that would be helpful:

- How to declare a separate environment for tests
- I am guessing by declaring a new ENV need a mechanism to authenticate a random test user
- After that I am hoping that if I can get the user authenticated, I can make normal post calls or service calls to test my app end to end.

Is there a better way other than running a separate test environment for the above purposes with a different config?

Another unrelated follow-up what is the best way to deal with many to many columns from another application?
Currently I do not see any fields in it. I have the following many to many relationship tables:
```sql
create table event_categories
(
    evententity_id      uuid           not null
        constraint fxyz
            references event,
    categoryentities_id numeric(19, 2) not null
);

create table categories_event
(
    category_id numeric(19, 2) not null,
    event_id    uuid           not null
        constraint fabc
            references event
);
```

In directus what I see in for this table is:
![image](https://user-images.githubusercontent.com/4995171/146669046-b31c5d01-1782-46fa-bfaf-452879846251.png)
![image](https://user-images.githubusercontent.com/4995171/146669054-3905503e-a15f-4372-b6dc-ae720e2f4357.png)

Is this some known problem or is there a better way of doing this?

Thanks!

**Comment by rijkvanzanten** [comment]
We don't have anything dedicated for testing extensions yet (cc @licitdev / @nickrum). Excited to hear any thoughts / suggestions you might have on the topic 🙂 

As for your other question; Directus currently requires tables to have a primary key in order to function. It looks like your junction tables don't currently have a primary key column

**Comment by gpietro** [comment]
A testing environment would be really helpful, especially for extensions like hooks, custom endpoints and the possiblity to write RBAC "scenarios".

**Comment by gpietro** [comment]
@rijkvanzanten is there a way to run a directus instance in test mode to execute unit tests with supertest?

Like Strapi does: https://docs.strapi.io/developer-docs/latest/guides/unit-testing.html#test-strapi-instance

Thanks

**Reply by benhaynes** [reply]
CC @licitdev

**Reply by keesvanbemmel** [reply]
Would be very interested in a testing setup for extensions as well!

**Reply by CherniyIrtish** [reply]
I also join to request.
